src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -5,248 +5,300 @@
+        /// - the deconstruction phase includes all the invocations of Deconstruct methods and tuple element accesses below a Deconstruct call
+        /// - the conversion phase
+        /// - the assignment phase
+        /// </summary>
+        private BoundExpression RewriteDeconstruction(BoundTupleExpression left, Conversion conversion, BoundExpression right)
         {
             var temps = ArrayBuilder<LocalSymbol>.GetInstance();
-            var stores = ArrayBuilder<BoundExpression>.GetInstance();
-            var placeholders = ArrayBuilder<BoundValuePlaceholderBase>.GetInstance();
-
-            // evaluate left-hand-side side-effects
-            ImmutableArray<BoundExpression> lhsTargets = GetAssignmentTargetsAndSideEffects(node.Left, temps, stores);
-
-            // get or make right-hand-side values
-            BoundExpression loweredRight = VisitExpression(node.Right);
-
-            ApplyDeconstructions(node, temps, stores, placeholders, loweredRight);
-            ApplyConversions(node, temps, stores, placeholders);
-            ApplyAssignments(node, stores, lhsTargets);
-
-            BoundExpression returnValue = MakeReturnValue(node, placeholders);
-            BoundExpression result = _factory.Sequence(temps.ToImmutable(), stores.ToImmutable(), returnValue);
-
-            RemovePlaceholderReplacements(placeholders);
-            placeholders.Free();
+            var effects = DeconstructionSideEffects.GetInstance();
+            ArrayBuilder<Binder.DeconstructionVariable> lhsTargets = GetAssignmentTargetsAndSideEffects(left, temps, effects.init);
 
-            temps.Free();
-            stores.Free();
+            BoundExpression returnTuple = ApplyDeconstructionConversion(lhsTargets, right, conversion, temps, effects, inInit: true);
+            if (!returnTuple.HasErrors)
+            {
+                returnTuple = VisitExpression(returnTuple);
+            }
+            BoundExpression result = _factory.Sequence(temps.ToImmutableAndFree(), effects.ToImmutableAndFree(), returnTuple);
+            Binder.DeconstructionVariable.FreeDeconstructionVariables(lhsTargets);
 
             return result;
         }
 
-        /// <summary>
-        /// Applies the deconstructions.
-        /// Adds any new locals to the temps and any new expressions to be evaluated to the stores.
-        /// </summary>
-        private void ApplyDeconstructions(BoundDeconstructionAssignmentOperator node, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> stores, ArrayBuilder<BoundValuePlaceholderBase> placeholders, BoundExpression loweredRight)
+        private BoundTupleLiteral ApplyDeconstructionConversion(ArrayBuilder<Binder.DeconstructionVariable> leftTargets,
+            BoundExpression right, Conversion conversion, ArrayBuilder<LocalSymbol> temps, DeconstructionSideEffects effects,
+            bool inInit)
```

Comment:
Any parts of the tree on the right which are above an invocation of `Deconstruct`.
So `(1, 2)` is in init. So is `e`.
But if `e` deconstructs into tuples or objects that can be further deconstructed, then any of the side-effects involved will go into the `effects.deconstructions` list, as opposed to the `effects.init` list.
I'll add a comment.